rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,745 @@
+  // and SaltLength must be provided as well.
+  //
+  // This rpc can be used to implement Go's [crypto.Signer] interface.
+  rpc Sign(SignRequest) returns (SignResponse) {}
+}
+
+message PublicKey {
+  // PublicKey is a public key encoded in PKIX, ASN.1 DER form.
+  bytes public_key_der = 1;
+}
+
+// SignRequest is a request to perform a signature with a specific agent key
+// with the provided signing parameters.
+message SignRequest {
+  // KeyRef refers to a specific key held by the agent.
+  KeyRef key_ref = 1;
+  // Digest is a hashed message to sign.
+  bytes digest = 2;
+  // HashName is the name of the hash used to generate the digest.
+  HashName hash_name = 3;
+  // SaltLength controls the length of the salt to be used in a PSS signature.
+  // Only required for PSS signatures.
+  SaltLength salt_length = 4;
+}
+
+// Signature is a private key signature.
+message Signature {
+  // For an RSA key, signature should be either a PKCS #1 v1.5 or PSS signature,
+  // depending on the hash and salt chosen. For an (EC)DSA key, it should be a
+  // DER-serialised, ASN.1 signature structure.
+  bytes signature = 1;
+}
+
+// KeyRef refers to a specific key.
+message KeyRef {
+  // ProxyHost is the root proxy hostname that a key is associated with.
+  string proxy_host = 1;
+  // Username is the username that a key is associated with.
+  string username = 2;
+  // KeyUsage is the key's usage.
+  KeyUsage key_usage = 3;
+}
+
+// KeyUsage specifies a key's usage.
+enum KeyUsage {
+  // Key type not specified.
+  KEY_TYPE_UNSPECIFIED = 0;
+  // Key type ssh.
+  KEY_TYPE_SSH = 1;
+  // Key type tls.
+  KEY_TYPE_TLS = 2;
+}
+
+// HashName is the name of a specific hashing algorithm.
+// These values match the Go [crypto] standard library.
+enum HashName {
+  HASH_NAME_UNSPECIFIED = 0;
+  HASH_NAME_MD4 = 1;
+  HASH_NAME_MD5 = 2;
+  HASH_NAME_SHA1 = 3;
+  HASH_NAME_SHA224 = 4;
+  HASH_NAME_SHA256 = 5;
+  HASH_NAME_SHA384 = 6;
+  HASH_NAME_SHA512 = 7;
+  HASH_NAME_MD5SHA1 = 8;
+  HASH_NAME_RIPEMD160 = 9;
+  HASH_NAME_SHA3_224 = 10;
+  HASH_NAME_SHA3_256 = 11;
+  HASH_NAME_SHA3_384 = 12;
+  HASH_NAME_SHA3_512 = 13;
+  HASH_NAME_SHA512_224 = 14;
+  HASH_NAME_SHA512_256 = 15;
+  HASH_NAME_BLAKE2s_256 = 16;
+  HASH_NAME_BLAKE2b_256 = 17;
+  HASH_NAME_BLAKE2b_384 = 18;
+  HASH_NAME_BLAKE2b_512 = 19;
+}
+
+enum SaltLength {
+  SALT_LENGTH_UNSPECIFIED = 0;
+  // Use the maximum length salt for a given hash function and public key.
+  SALT_LENGTH_MAX = 1;
+  // Use a salt equal in length to the chosen hash used.
+  SALT_LENGTH_HASH_LENGTH = 2;
+}
+```
+
+#### Teleport Connect `TerminalService`
+
+```diff
+### teleterm/v1/service.proto
+service TerminalService {
+  ...
++  // StartKeyAgent starts the Teleport Key Agent.
++  rpc StartKeyAgent(google.protobuf.Empty) returns (google.protobuf.Empty);
++  // StopKeyAgent stops the Teleport Key Agent.
++  rpc StopKeyAgent(google.protobuf.Empty) returns (google.protobuf.Empty);
```

Comment:
Please use named request and response messages. Changing types from Empty to a named message is a breaking change (if we ever need to).